### PR TITLE
Hide commands from ApplicationDescriptor, but allow invoking

### PR DESCRIPTION
**Pull request by jwdeitch**
I would like to hide commands from cluttering the descriptors, but still allow their invocation from code or cron.

| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | comma-separated list of tickets fixed by the PR, if any |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

### CODE REVIEW

File: src/Symfony/Component/Console/Command/Command.php [hunk 1]
```diff
@@ -34,6 +34,7 @@ class Command
     private $processTitle;
     private $aliases = array();
     private $definition;
+    private $hidden;
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Command/Command.php]: Should be `$this->hidden = false;`

File: src/Symfony/Component/Console/Command/Command.php [hunk 2]
```diff
@@ -447,6 +448,30 @@ public function getName()
     }
 
     /**
+     * Sets if the command should be hidden from application inspection.
+     *
+     * @param bool $hiddenBool To show this command or not
+     * 
+     * @return Command The current instance
+     */
+    public function setHidden($hiddenBool)
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Command/Command.php]: Should be `$hidden`

Comment (jwdeitch) [inline src/Symfony/Component/Console/Command/Command.php]: indeed. Thank you sir

File: src/Symfony/Component/Console/Command/Command.php [hunk 3]
```diff
@@ -447,6 +448,30 @@ public function getName()
     }
 
     /**
+     * Sets if the command should be hidden from application inspection.
+     *
+     * @param bool $hiddenBool To show this command or not
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Command/Command.php]: Derp.

Comment (ro0NL) [inline src/Symfony/Component/Console/Command/Command.php]: What about `To hide the command in application descriptions`.

**Comment by ogizanagi** [comment]
What about other descriptors ?

**Comment by ro0NL** [comment]
@ogizanagi all format use `ApplicationDescription::getCommands`. https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Descriptor/JsonDescriptor.php#L67

**Comment by jwdeitch** [comment]
@ogizanagi Sorry which speicifically? The other descriptors seems to call `ApplicationDescription->getCommands()`

**Comment by ogizanagi** [comment]
@ro0NL, @jwdeitch : Oh right. Sorry. PR title & description may be updated I guess then 😉 .

**Comment by javiereguiluz** [comment]
Could we change the method names to follow the same notation as for services? Instead of `setHidden()` and `isHidden()` -> `setPublic(true / false)` and `isPublic()` ?

**Comment by ro0NL** [comment]
Maybe if we run a private command it should give you a heads up in verbose mode..  would that be useful?

**Comment by jwdeitch** [comment]
Good idea. Should we merge this first then address that? Or I can throw it into this PR

**Comment by ogizanagi** [comment]
@jwdeitch : Could you update the testsuite (`AbstractDescriptorTest`) by adding a new private command in `DescriptorApplication2` ?

File: src/Symfony/Component/Console/Command/Command.php [hunk 4]
```diff
@@ -447,6 +448,30 @@ public function getName()
     }
 
     /**
+     * To hide the command in application descriptions.
+     *
+     * @param bool $public To show this command or not
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Command/Command.php]: I'd suggest something like: `Whether the command should be publicly shown or not.` and remove the above description (which brings nothing more IMHO).

File: src/Symfony/Component/Console/Command/Command.php [hunk 5]
```diff
@@ -447,6 +448,30 @@ public function getName()
...
+    {
+        $this->public = $public;
+
+        return $this;
+    }
+
+    /**
+     * Returns if the command is public in application descriptions.
+     *
+     * @return bool If the command is public or not
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Command/Command.php]: Same. I feel this repetition is useless.

**Review by ogizanagi — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Console/Command/Command.php [hunk 6]
```diff
@@ -447,6 +448,26 @@ public function getName()
     }
 
     /**
+     * @param bool $public Whether the command should be publicly shown or not.
+     * 
+     * @return Command The current instance
+     */
+    public function setPublic($public)
+    {
+        $this->public = $public;
```
Comment (fabpot) [inline src/Symfony/Component/Console/Command/Command.php]: should be `$this->public = (bool) $public;`

File: src/Symfony/Component/Console/Command/Command.php [hunk 7]
```diff
@@ -447,6 +448,26 @@ public function getName()
     }
 
     /**
+     * @param bool $public Whether the command should be publicly shown or not.
+     * 
+     * @return Command The current instance
+     */
+    public function setPublic($public)
```
Comment (fabpot) [inline src/Symfony/Component/Console/Command/Command.php]: What about `setPrivate()` without an argument? I don't see a need to switch from private to public.

Comment (javiereguiluz) [inline src/Symfony/Component/Console/Command/Command.php]: I suggested changing it to `setPublic(false)` to mimic Symfony services ... where you must set `public: false` for private services. If we use opposing techniques to do the same (`public: false` for services, `private: true` for commands) we're increasing the learning curve.

Comment (jwdeitch) [inline src/Symfony/Component/Console/Command/Command.php]: we could meet in the middle at `setPublicity` as it's not favoring one way or another :)

**Comment by fabpot** [comment]
Thank you @jwdeitch.

**Comment by wouterj** [comment]
I think calling this *private* will have the same confusion that was just solved with the services: I expect I cannot run private services outside the app, but I still can. It's simply only hidden in the command listing.

Calling this *hidden* does a better job in describing the feature imo.

**Comment by fabpot** [comment]
I tend to agree with @wouterj. A private command is indeed not really private in the same sense as for the container, it's just hidden from the list of commands and can still be executed like any other commands. @javiereguiluz WDYT?

**Comment by javiereguiluz** [comment]
I understand "private" as something hidden, secret, confidential, special. Those are precisely the [synonyms of the "private" word](http://www.thesaurus.com/browse/private). If we talk about commands that cannot be executed outside the app, to me that would be: "banned", "secured", "disabled", "closed", etc.

But if this is confusing to others, then "hidden" is also a very understandable name for this feature.

**Comment by stof** [comment]
@javiereguiluz the confusion is that it has a different behavior than ``private`` in the DI component:

- in DI, a private service can be injected into other services, but cannot be accessed from the outside of the container
- in console, a private command can be called like any other command, as long as you know it exists.

**Comment by xabbuh** [comment]
I agree with @wouterj and @fabpot. I opened #20266 which would rename `$public` to `$hidden`.

**Pull request merged**